### PR TITLE
Removed auto-decoding from std.conv.parse (part 2)

### PR DESCRIPTION
**Pull request by JackStouffer**
Benchmark code here: https://gist.github.com/JackStouffer/af53d497532a864f56a5b53cb35cd438

Gives about a 40% speed increase:

```
# old
$ ldc2 -O5 -release -boundscheck=off test.d && ./test
Total time: 5 secs, 266 ms, 64 μs, and 3 hnsecs

# new
$ ldc2 -O5 -release -boundscheck=off test.d && ./test
Total time: 3 secs, 295 ms, 663 μs, and 7 hnsecs
```

### CODE REVIEW

**Comment by JackStouffer** [comment]
Ping @9il

**Comment by 9il** [comment]
LGTM

**Comment by 9il** [comment]
Please add small msg to change log

**Comment by JackStouffer** [comment]
@9il Done

**Comment by 9il** [comment]
@JackStouffer  please add test with global `__gshared char[]` integer number and `pragma(inline, false)` for parse functions. 4x-5x is large number, so,  we need to be sure that LDC optimisation does not brake the test.

**Comment by wilzbach** [comment]
> @JackStouffer please add test with global __gshared char[] integer number and pragma(inline, false) for parse functions. 4x-5x is large number, so, we need to be sure that LDC optimisation does not brake the test.

IIRC @klickverbot pointed out that a good way to prevent LDC optimizations is to use separate files & annotate the tested methods with `extern`.

I have a [small repo](https://github.com/wilzbach/d-benchmarks) for benchmark scripts which are separately compiled and then linked, maybe that helps you?

**Comment by 9il** [comment]
btw, `pragma(inline, false)` works correctly only in the recent LDC alpha, or master

**Comment by JackStouffer** [comment]
Rebased.

Will add extra test when I get time.

**Comment by JackStouffer** [comment]
@9il I updated the benchmark code here to your requests: https://github.com/dlang/phobos/pull/4597

**Comment by JackStouffer** [comment]
Ping @9il

**Comment by andralex** [comment]
This doesn't look right. Most code should stay the same. The only change is at the top level, where decodable strings should be replaced with `byChar` or `.representation`.

**Comment by andralex** [comment]
Where is part 1? I hope it doesn't make similar changes... :)

**Comment by JackStouffer** [comment]
@andralex Here's the problem, as outlined [here](https://github.com/dlang/phobos/pull/4597)

> The reason why this is so messy is parse takes its range by ref, so byCodeUnit cannot be used without using refRange, which requires using &r, which is unsafe. So the only solution left is to do it manually with a bunch of static ifs.

> I tried putting the refRange.byCodeUnit in a @trusted block, but that resulted in a speed pessimization (I'm guessing due to the inability to inline at that point).

**Comment by andralex** [comment]
@JackStouffer then just work on a temporary and in the end assign the reference from it. It's a classic.

**Comment by JackStouffer** [comment]
I don't understand. Operate on a temporary and reassign it how?

If I make a copy of `p` inside of `parse` and then use `byCodeUnit` on that, the copy of `p` is copied again by `byCodeUnit` and that is advanced and not the copy of `p`. So I would still have to manually `popFront` the copy in order to reassign it to `p`.

**Comment by JackStouffer** [comment]
Ping @andralex @9il

File: changelog.dd [hunk 1]
```diff
@@ -71,6 +71,9 @@ $(BUGSTITLE Library Changes,
     $(LI $(RELATIVE_LINK2 emplace-inner-class, `std.conv.emplace` no longer allows
     to emplace classes directly nested inside other classes without specifying a
     suitable `outer` pointer))
+    $(LI Conversions from strings to number types using $(REF to, std,conv)
+        and $(REF parse, std,conv) were optimized. On average, integer types
+        are three times faster, and floating point types are 40% faster.)
```
Comment (9il) [inline changelog.dd]: change log should be updated. ping @wilzbach . You wanted to change current change log system

**Comment by 9il** [comment]
@JackStouffer it is save to reassing ubyte array to string after intefer parsing. Please implement Andrei's suggestions

**Comment by JackStouffer** [comment]
I will get back to this eventually. I'm a bit swamped with school and work right now, so maybe around Thanksgiving.

**Comment by andralex** [comment]
@JackStouffer thanks, ping me when ready.

**Comment by JackStouffer** [comment]
Closed in favor of https://github.com/dlang/phobos/pull/5015

**Pull request closed without merging**